pkg/agent/core/ngt/handler/grpc/object.go

```diff
@@ -241,27 +241,34 @@ func (s *server) StreamListObject(_ *payload.Object_List_Request, stream vald.Ob
 		mu.Unlock()
 
 		if err != nil {
-			emu.Lock()
-			errs = append(errs, err)
-			emu.Unlock()
+			emu.RLock()
+			_, ok := emap[err.Error()]
+			emu.RUnlock()
+			if !ok {
+				emu.Lock()
+				errs = append(errs, err)
+				emap[err.Error()] = struct{}{}
+				emu.Unlock()
+			}
 		}
 
 		// always return true to continue streaming and let the context cancel the Range when stream closes.
 		return true
 	})
 
 	if len(errs) != 0 {
-		slices.RemoveDuplicates(errs, func(left, right error) bool {
-			return left.Error() < right.Error()
-		})
-		err = errors.Join(errs...)
-		st, msg, err := status.ParseError(err, codes.Internal, "failed to parse StreamListObject final gRPC error response")
-		if span != nil {
-			span.RecordError(err)
-			span.SetAttributes(trace.FromGRPCStatus(st.Code(), msg)...)
-			span.SetStatus(trace.StatusError, msg)
+		// Register all the gRPC codes to the span. Doing this because ParseError cannot parse joined error.
+		for _, e := range errs {
+			st, msg, err := status.ParseError(e, codes.Internal, "failed to parse StreamListObject final gRPC error response")
+			if span != nil {
+				span.RecordError(err)
+				span.SetAttributes(trace.FromGRPCStatus(st.Code(), msg)...)
+				span.SetStatus(trace.StatusError, msg)
+			}
 		}
-		return err
+
```

Comment:
seems we can rewrite it like
```
if span != nil {
	for _, e := range errs {
		st, msg, err := status.ParseError(e, codes.Internal, "failed to parse StreamListObject final gRPC error response")
		span.RecordError(err)
		span.SetAttributes(trace.FromGRPCStatus(st.Code(), msg)...)
		span.SetStatus(trace.StatusError, msg)
	}
}
```
because this block of logic only effective when span is not nil